fix: Extract Gator Permission Expiry From Delegation Context

hanzel98 opened the pull request:
## **Description**

Fixes #464 - Gator permissions with expiry were displaying "No expiration" in the UI.

**Root Cause:** The expiry timestamp was excluded as part of a sanitization of the permissions response because it was stored in the permission. 

**Solution:** 
- Include the rules as part of the permissions responde so we can find the expirty inside, this part of the work is done in the gator-permissions-controller

## **Changelog**

CHANGELOG entry: Fixed gator permissions displaying "No expiration" when expiry timestamp exists

## **Related issues**

Fixes: #464

## **Manual testing steps**

1. Create a gator permission with an expiry timestamp from `jeff.dripfund.xyz`
2. Navigate to Settings → Permissions → All Permissions
3. Locate the gator permission from `jeff.dripfund.xyz`
4. Expand the permission details
5. Verify the expiration date displays correctly (e.g., "11/13/2026") instead of "No expiration"

## **Screenshots/Recordings**

### **Before**
Permission showed "No expiration" despite having an expiry timestamp in the delegation context.

### **After**
Permission now correctly displays the expiration date (e.g., "11/13/2026") extracted from the `TimestampEnforcer` caveat.

## **Pre-merge author checklist**

- [x] I've followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Extension Coding Standards](https://github.com/MetaMask/metamask-extension/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I've included tests if applicable (7 comprehensive test cases covering all edge cases)
- [x] I've documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [ ] I've applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [x] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.


<img width="1278" height="2074" alt="image" src="https://github.com/user-attachments/assets/b5e29995-4127-49b1-bac5-c6e49a73dd41" />



<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> Fixes incorrect "No expiration" display by sourcing expiry from delegation rules.
> 
> - Use `extractExpiryToReadableDate(permissionResponse.rules || [])` in `ReviewGatorPermissionItem` for stream/periodic/revocation permissions; fallback to `gatorPermissionNoExpiration` when no expiry rule exists
> - Update memo/callback dependencies to include `permissionResponse.rules`
> - Tests: add expiry rule to mocks and assert exact dates; add cases for no rules and non-expiry rules showing "No expiration"; cover native and ERC20 stream/periodic, revocation, and date formatting
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 464f0b5e46da88ee17ae135f7a542c5feb4b432e. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->





Depends on this one https://github.com/MetaMask/core/pull/7195 Pending to upgrade to the gator-permissions-controller: ^0.6.0

hanzel98 committed:
fix: Extract gator permission expiry from delegation context

Fixes #464 - Permission with expiry showing 'No expiration'

- Extract expiry timestamp from permissionResponse.context by decoding delegation
- Find TimestampEnforcer caveat and decode the terms field (last 32 hex chars)
- Convert BigInt timestamp to readable date format
- Add comprehensive test coverage for all edge cases:
  - Valid expiry extraction
  - No TimestampEnforcer caveat
  - Invalid delegation count (0 or >1)
  - Invalid terms length
  - Zero timestamp
  - Decoding errors
- Remove unused imports and debug logs

hanzel98 committed:
fix: configure mocks for expiration date extraction in ReviewGatorPermissionItem tests

- Add beforeEach hooks to configure decodeDelegations and getDeleGatorEnvironment mocks for NATIVE and ERC20 token test suites
- Update test assertions to verify actual expiration date content instead of just element presence
- Fix variable naming conflicts in edge case tests to avoid scope shadowing
- Ensures tests correctly validate expiration date extraction from delegation context

MoMannn commented on ui/components/multichain/pages/gator-permissions/components/review-gator-permission-item.tsx:
```diff
@@ -164,23 +162,71 @@ export const ReviewGatorPermissionItem = ({
-      if (!rules) {
+  const getExpirationDate = useCallback((): string => {
+    try {
+      const delegations = decodeDelegations(permissionContext);
```
This looks like it might be usable in different places. Could we wrap it into a reusable function that returns expiry or 0 and move it to: `gator-permissions-utils` file.

MoMannn reviewed (changes requested):
Some tests are failing. Run `yarn test:unit:coverage --shard=4/6` locally to check.

hanzel98 committed:
test: fixed tests

hanzel98 committed:
refactor: extract expiry timestamp logic to reusable utility function

- Created extractExpiryTimestampFromDelegation() in time-utils.ts
- Returns expiry timestamp or 0, making it reusable across components
- Simplified review-gator-permission-item.tsx from ~60 lines to ~10 lines
- Moved comprehensive tests from component to utility tests
- Component tests now focus on UI behavior only

hanzel98 committed:
fix: remove invalid timeout parameters from test cases

Jest's it() function doesn't accept timeout as third parameter with async callbacks.
Removed invalid timeout arguments from two test cases that were causing TypeScript errors.

hanzel98 committed:
fix: remove invalid timeout parameters from beforeAll/afterAll hooks

Jest's beforeAll/afterAll don't accept timeout as second parameter with async callbacks.
Removed invalid timeout arguments that were causing issues.

hanzel98 committed:
chore: deleted utc assumption

hanzel98 committed:
refactor: migrate extractExpiryTimestampFromDelegation to gator-permissions-controller

hanzel98 committed:
chore: extract expiration from rules in permissions response

hanzel98 committed:
refactor: simplify expiry extraction using extractExpiryToReadableDate utility

- Use extractExpiryToReadableDate from time-utils instead of manual extraction
- Remove UTC zone specification from convertTimestampToReadableDate
- Reduce getExpirationDate function from 26 to 9 lines
- Import GatorPermissionRule type for proper typing

hanzel98 committed:
Merge branch 'main' into chore/expiration-date-fixed

# Conflicts:
#	package.json
#	ui/components/multichain/pages/gator-permissions/components/review-gator-permission-item.tsx
#	yarn.lock

hanzel98 committed:
fix: Pass rules to getExpirationDate for all permission types

- Fix stream permissions to pass permissionResponse.rules to getExpirationDate
- Fix periodic permissions to pass permissionResponse.rules to getExpirationDate
- Fix revocation permissions to use permissionResponse.rules instead of permissionResponse.permission.rules
- Update dependency arrays to include permissionResponse.rules for proper re-rendering

This fixes the issue where expiration dates were not displaying because getExpirationDate was called without the required rules parameter.

hanzel98 committed:
fix: Replace mockSelectedAccountAddress with mockAccountAddress in test file

- Fix TS2304 errors on lines 733 and 785
- Use correct variable name that is defined in the file

MoMannn committed:
Merge branch 'main' into chore/expiration-date-fixed

MoMannn committed:
Show no expiration if no expity rule

MoMannn committed:
Merge branch 'main' into chore/expiration-date-fixed
